### DISCUSSION TITLE
mapValuesDeep mutates the original object if `leavesOnly: false`

### DISCUSSION THREAD

**Discussion by lewisblackwood** [YuriGor/deepdash#100] [Q&A]
Hi,

Great library 👍 

I found using `mapValuesDeep` that if the options are left at the default `leavesOnly: false`, then it mutates the original object. If `leavesOnly: true` then the original object is not mutated.

You can see an example repl.it [here](https://replit.com/@lewisblackwood/mapValuesDeep).

```javascript
// leavesOnly: true
{ sourceData: { hello: 'world' }, mappedData: { hello: 'WORLD' } }
// leavesOnly: false
{ sourceData: { hello: 'WORLD' }, mappedData: { hello: 'WORLD' } }
```

Is this the expected behaviour?

**Comment by YuriGor** [comment]
Hi, thank you for report, looks like a bug, will check in next few days.

**Comment by YuriGor** [comment]
If `leavesOnly` is false, it will call your callback for middle object values also, not only  for primitive/childless 'leaves'.
So you return original object `v` here:
`const iteratee = (v, k) => k === "hello" ? v.toUpperCase() : v`
And it's placed to the result as provided by you.
So in result object you have same references as in source.

After that deepdash goes deeper and asks about leaf string value, but this leaf already belongs both to source and result objects, because they were mixed on previous iteration.
So it's not `mapValuesDeep` mutates the original object, it's you :)
If you want to go over non-leaf nodes (from your example it's not clear why do you need it) - don't return original nodes by reference, always replace them by something or clone.
